Wait for wsl-init to finish before moving on with wsl setup.

vladonemo opened the pull request:
Partially fixes . #1919 

---
Edit by @mook-as: Disable GitHub auto-linking of issues — this isn't necessarily going to fix all of #1919

mook-as reviewed (changes requested):
Is my understanding correct that this is trying to fix an issue were `/var/lib` / `/etc/rancher` are not staying mounted for `wsl-init`?

If that is the case, I think this fix is not all the way correct: the desired flow is that `this.mountData()` (line 1203) mounts things, and `wsl-init` sees it.  However, I guess it's possible that the machine is busy enough that none of the other WSL processes start before all the WSL processes in `this.mountData()` exits (expected) _and_ WSL times out the idle distro and terminates it (unexpected).  In this case, I believe a proper fix would be:
- Make `this.mountData()` spawn a process that stays running (to prevent the distro from shutting down); presumably this would have to be returned (as a [`ChildProcess`](https://nodejs.org/docs/latest-v16.x/api/child_process.html#class-childprocess))
- Run `wsl-init` and friends, as we do now.
- Terminate the process we got from the first step; at this point `wsl-init` is actively running so the VM still can't auto-terminate.

Thanks for your PR! However, it doesn't appear to actually eliminate the cause (unless I'm misunderstanding what the cause is, in which case please let me know).

vladonemo commented on src/k8s-engine/wsl.ts:
```diff
@@ -1201,7 +1201,8 @@ export default class WSLBackend extends events.EventEmitter implements K8s.Kuber
         await this.progressTracker.action(this.lastCommandComment, 100, this.mountData());
         this.lastCommandComment = 'Starting WSL environment';
-        const installerActions = [
+6
```
oh, my bad.

vladonemo commented on src/k8s-engine/wsl.ts:
```diff
@@ -1201,7 +1201,8 @@ export default class WSLBackend extends events.EventEmitter implements K8s.Kuber
         this.lastCommandComment = 'Starting WSL environment';
-        const installerActions = [
+6
+        await Promise.all([
           this.progressTracker.action(this.lastCommandComment, 100, async() => {
```
right, haven't realized it's just a single one. thanks

vladonemo commented:
> Is my understanding correct that this is trying to fix an issue were `/var/lib` / `/etc/rancher` are not staying mounted for `wsl-init`?
> 
> If that is the case, I think this fix is not all the way correct: the desired flow is that `this.mountData()` (line 1203) mounts things, and `wsl-init` sees it. However, I guess it's possible that the machine is busy enough that none of the other WSL processes start before all the WSL processes in `this.mountData()` exits (expected) _and_ WSL times out the idle distro and terminates it (unexpected). In this case, I believe a proper fix would be:
> 
> * Make `this.mountData()` spawn a process that stays running (to prevent the distro from shutting down); presumably this would have to be returned (as a [`ChildProcess`](https://nodejs.org/docs/latest-v16.x/api/child_process.html#class-childprocess))
> * Run `wsl-init` and friends, as we do now.
> * Terminate the process we got from the first step; at this point `wsl-init` is actively running so the VM still can't auto-terminate.
> 
> Thanks for your PR! However, it doesn't appear to actually eliminate the cause (unless I'm misunderstanding what the cause is, in which case please let me know).

Could you please elaborate on the 'WSL times out the idle distro and terminates it'? I don't know this feature of WSL. 

The code awaits until the `mountData` finishes fully. Then it starts the WSL env (the whole block with `runInit` at the end). And only after this is done (awaiting the `wsl-init.pid`) the installation continues. I must admit, I still don't fully understand the actual root cause, but I believe that running (one of the) installations in parallel to the `runInit` leads to the race condition.

vladonemo commented on src/k8s-engine/wsl.ts:
```diff
@@ -1233,23 +1234,23 @@ export default class WSLBackend extends events.EventEmitter implements K8s.Kuber
-              await this.persistVersion(actualDesiredVersion);
-            })
-          );
+          await this.progressTracker.action('Installing k3s', 100, async() => {
```
right, will do so

mook-as commented:
Hi!
I'm referring to [this bit in the WSL documentation](https://docs.microsoft.com/en-us/windows/wsl/wsl-config#the-8-second-rule), where it describes how WSL will terminate a distribution after about 8 seconds. We're actually kind of depending on this right now, since `mountData()` fully exists, and we expect `wsl-init` to see the same mounts from there (even though WSL could have terminated the distribution within that window between the two function calls, losing the mount information).

I'm suggesting keeping a (WSL) process around across that gap, so that WSL cannot terminate our distribution before `wsl-init` start running.

vladonemo commented:
Ah, ok, now I see. You mean the 8 seconds rule after approximately which the distro is stopped. Thanks for the explanation. I understand it now. Will implement it according to your suggestion.

vladonemo commented:
I reverted all my previous changes and implemented as per your suggestion. The `mountData` now spawns a child process (plain `sh`) which is killed after the `runInit` finishes. Tested it and it starts OK now ;)

mook-as commented on src/k8s-engine/wsl.ts:
```diff
@@ -1233,6 +1238,7 @@ export default class WSLBackend extends events.EventEmitter implements K8s.Kuber
             await this.execCommand('mkdir', '-p', '/var/lib/misc');
 
             await this.runInit();
+            lock.kill('SIGTERM');
```
This should always run, so there should be a `try/finally` somewhere.

I thinking something like
```ts
try {
  await Promise.all(installerActions);
} finally {
  lock.kill('SIGTERM');
}
```
Note that this code is Windows-only, so [according to docs](https://nodejs.org/docs/latest-v16.x/api/child_process.html#subprocesskillsignal) the arugment here doesn't matter and the process is always killed with a `TerminateProcess`.

mook-as commented on src/k8s-engine/wsl.ts:
```diff
@@ -1575,8 +1581,10 @@ export default class WSLBackend extends events.EventEmitter implements K8s.Kuber
     await this.progressTracker.action(this.lastCommandComment, 5, async() => {
       await this.stop();
       // Mount the data first so they can be deleted correctly.
-      await this.mountData();
+      const lock = await this.mountData();
+
       await this.k3sHelper.deleteKubeState((...args) => this.execCommand(...args));
+      lock.kill('SIGTERM');
```
See above — `try/finally` here too please.

mook-as reviewed (changes requested):
Thanks for the changes! Just some smaller fixes now.

mook-as commented:
Hi! We're trying to land this for an impending release; I've pushed the changes I want to your branch (on top of your commits) and I'll arrange for a review shortly.

Thanks!

rak-phillip reviewed (approved):
Changes look good and pass smoke tests 🙂
